### PR TITLE
JAMES-2019 Cannot preview because it starts with white space strings

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java [hunk 1]
```diff
@@ -41,8 +43,9 @@ public ParsedContent extractContent(InputStream inputStream, String contentType,
             return new ParsedContent(IOUtils.toString(inputStream), emptyMetadata);
            }
            if (contentType.equals("text/html")) {
-               String text = Jsoup.parse(IOUtils.toString(inputStream)).text();
-               return new ParsedContent(text, emptyMetadata);
+               Document doc = Jsoup.parse(IOUtils.toString(inputStream));
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: WARNING: Where is the encoding?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package org.apache.james.mailbox.inmemory;
```
Comment (chibenwa) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java]: License

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java [hunk 3]
```diff
@@ -41,8 +43,9 @@ public ParsedContent extractContent(InputStream inputStream, String contentType,
...
+               doc.select(TITLE_HTML_TAG).remove();
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: Why removing this HTML tag?

File: mailbox/memory/src/test/resources/documents/html.txt [hunk 4]
```diff
@@ -0,0 +1,1110 @@
+<!doctype html>
```
Comment (chibenwa) [inline mailbox/memory/src/test/resources/documents/html.txt]: License in a HTML comment?

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java (hunk 3 above)
Comment (quynhn) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: - First of all to make it's consistence with Tika, Tika does not contain <title> tag in extracted text. And we have only one IT for both Memory and Cassandra. Of course that we can separated 2 ITs. But the other.
- The second reason, I think it makes sense and all the content does not belong to <body> tag should not belong to extracted text.

**Comment by chibenwa** [comment]
It's cool...

But it's not what I asked for...

If I receive this text/plain email:

```
                                   Hello
                         I
                                     love
                        spaces
                     !!!!!!!!!!!!!!!!!!!!!
```

Could you ensure I have this summary?

```
Hello I love spaces!!!!!!!!!!!!!!!!
```

(Basically trim before keeping 256 char)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 5]
```diff
@@ -228,6 +228,11 @@ public void appendMessageWithSpecificCharset(String messageName, String mailbox)
         appendMessage(messageName, "eml/windows1252charset.eml");
```
Comment (strokyl) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: Typo in the commit s/intergration/integration/g

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreviewGenerator.java [hunk 6]
```diff
@@ -35,6 +35,6 @@ public String compute(Optional<String> textBody) {
     }
 
     private String truncateToMaxLength(String body) {
-        return StringUtils.left(body, MAX_PREVIEW_LENGTH);
+        return StringUtils.left(StringUtils.normalizeSpace(body), MAX_PREVIEW_LENGTH);
```
Comment (strokyl) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreviewGenerator.java]: Where is the associated test ? If it's in a previous or next commit I think you should merge those commit

**Comment by chibenwa** [comment]
Could you add a test proving that you are normalising spaces?

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
... à son travail et ses valeurs. Ce partenariat et cette initiative vont dans le sens de notre objectif principal de LINAGORA Vietnam : contribuer à l'indépendance numérique du Vietnam via la fourniture des solutions d'e-Gouvernance Open Source aux administrations vietnamiennes et le développement des programmes de formation de qualité sur l'informatique et les technologies Open Source. Share Tweet Share En savoir plus sur Passerelles Numériques : Passerelles Numériques est une ONG française. Ils opèrent au Cambodge, aux Philippines et au Vietnam. Leur mission est de permettre à des jeunes très défavorisés d’accéder à une éducation et à une formation technique et professionnelle, dans le secteur du numérique. Cela leur permettra d’échapper durablement à la pauvreté, et de contribuer au développement socio-économique de leur pays. Contact presse : Robin Ledru : communication@linagora.com et 06 33 87 88 64 Suivez nous sur nos différents réseaux sociaux Copyright © 2017 *|LIST:COMPANY|*, All rights reserved. *|LIST:DESCRIPTION|* Our mailing address is: *|HTML:LIST_ADDRESS_HTML|* Want to change how you receive these emails? You can update your preferences or unsubscribe from this list");
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java]: equality is not the right solution here

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    @Before
+    public void setUp() {
+        textExtractor = new JsoupTextExtractor();
+    }
+
+    @Test
+    public void HtmlTextTest() throws Exception {
+        InputStream inputStream = ClassLoader.getSystemResourceAsStream("documents/html.txt");
+        assertThat(inputStream).isNotNull();
+        assertThat(textExtractor.extractContent(inputStream, "text/html", null).getTextualContent())
```
Comment (mbaechler) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java]: what does the null stands for

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java (hunk 3 above)
Comment (mbaechler) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: does it remove the tag or the tag + the title content ?

Comment (quynhn) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: It removes fully the tag (tag and its content)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreviewGenerator.java (hunk 6 above)
Comment (quynhn) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreviewGenerator.java]: It's missed then I add one more

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java (hunk 8 above)
Comment (quynhn) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java]: It's there for sure that the resource is existed. But to make only one assertion, I will remove it

File: mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java (hunk 7 above)
Comment (quynhn) [inline mailbox/memory/src/test/java/org/apache/james/mailbox/inmemory/JsoupTextExtractorTest.java]: I changed

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java [hunk 9]
```diff
@@ -27,22 +27,26 @@
...
 import org.apache.james.mailbox.extractor.TextExtractor;
 import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
 
 import com.google.common.collect.Maps;
 
 
 public class JsoupTextExtractor implements TextExtractor {
+    private static final String TITLE_HTML_TAG = "title";
+    private static final String UTF8_ENCODING = "UTF-8";
```
Comment (mbaechler) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: UTF-8 constant is already defined in guava Charsets

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 10]
```diff
@@ -285,3 +285,10 @@ Feature: GetMessages method
     And the preview of the message is "àààà éééé èèèè"
     And the textBody of the message is "àààà\r\n\r\néééé\r\n\r\nèèèè\r\n"
     And the htmlBody of the message is "<html>\r\n  <p>àààà</p>\r\n  <p>éééé</p>\r\n  <p>èèèè</p>\r\n</html>\r\n"
+
+  Scenario: Preview should respect in special cases
+    Given the user has a message "m1" in "INBOX" mailbox with special case
+    When the user ask for messages "m1"
+    Then no error is returned
+    And the list should contain 1 message
+    And the preview of the message is "Afficher l'email dans le navigateur. Hackathon sur les technologies Open Source avec Passerelles Numériques à Danang LINAGORA est fière d'être partenaire de Passerelles Numériques pour animer un Hackathon avec des jeunes étudiants dynamiques et prometteurs"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: strict equality is too much

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 11]
```diff
@@ -285,3 +285,10 @@ Feature: GetMessages method
     And the preview of the message is "àààà éééé èèèè"
     And the textBody of the message is "àààà\r\n\r\néééé\r\n\r\nèèèè\r\n"
     And the htmlBody of the message is "<html>\r\n  <p>àààà</p>\r\n  <p>éééé</p>\r\n  <p>èèèè</p>\r\n</html>\r\n"
+
+  Scenario: Preview should respect in special cases
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: please define "special case"

File: mailbox/memory/src/test/resources/documents/html.txt (hunk 4 above)
Comment (strokyl) [inline mailbox/memory/src/test/resources/documents/html.txt]: Why not, don't you considere HTML as code :trollface:

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java [hunk 12]
```diff
@@ -64,4 +64,12 @@ public void computeShouldReturnStringIsLimitedTo256Length() throws Exception {
             .hasSize(MessagePreviewGenerator.MAX_PREVIEW_LENGTH)
             .isEqualTo(expected);
     }
+
+    @Test
+    public void computeShouldReturnNormalizeSpaceString() throws Exception {
```
Comment (strokyl) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java]: What about tabs and other special space ?

Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java]: I used the StringUtils function which defines 
`S (white space) consists of one or more space (#x20) characters, carriage returns, line feeds, or tabs.
White Space
[3]   	S	   ::=   	(#x20 | #x9 | #xD | #xA)+`

**Comment by quynhn** [comment]
> Why not, don't you considere HTML as code

To make it more confidence, I got the html content exactly from email :)

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java (hunk 9 above)
Comment (quynhn) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/JsoupTextExtractor.java]: Yes

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 11 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: I changed a bit

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java (hunk 12 above)
Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java]: Could you have a small test proving it?

I don't like going to read other project documentation, and I don't trust them.... :-)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 10 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: I changed to preview should not contain some white spaces characters. I don't know another way to assert this case

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 13]
```diff
@@ -285,3 +285,10 @@ Feature: GetMessages method
     And the preview of the message is "àààà éééé èèèè"
     And the textBody of the message is "àààà\r\n\r\néééé\r\n\r\nèèèè\r\n"
     And the htmlBody of the message is "<html>\r\n  <p>àààà</p>\r\n  <p>éééé</p>\r\n  <p>èèèè</p>\r\n</html>\r\n"
+
+  Scenario: Preview should be normalized in case of long and complicated HTML content
+    Given the user has a message "m1" in "INBOX" mailbox with long and complicated HTML content
+    When the user ask for messages "m1"
+    Then no error is returned
+    And the list should contain 1 message
+    And the preview of the message is present and normalized
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: the problem is : the steps of your test should help understand what the title means. In this case, anybody reading the test case should be able to learn what "normalized" means.
What about : "the preview should not contain consecutive spaces or blank characters" and "the preview should not be empty"

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java (hunk 12 above)
Comment (quynhn) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MessagePreviewGeneratorTest.java]: I changed the input for that which contain all white space. Is that enough or does it need to separate them all?

**Pull request closed without merging**